astropy/coordinates/representation.py

```diff
@@ -1593,6 +1618,55 @@ def represent_as(self, other_class, differential_class=None):
 
         return super().represent_as(other_class, differential_class)
 
+    def transform(self, matrix):
+        r"""Transform the unit-spherical coordinates using a 3x3 matrix.
+
+        This returns a new representation and does not modify the original one.
+        Any differentials attached to this representation will also be
+        transformed.
+
+        Parameters
+        ----------
+        matrix : (3,3) array-like
+            A 3x3 transformation matrix, such as a rotation matrix.
+
+        Returns
+        -------
+        `UnitSphericalRepresentation` or `SphericalRepresentation`
+            If ``matrix`` is O(3) -- :math:`M \dot M^T = I` -- like a rotation,
+            then the result is a `UnitSphericalRepresentation`.
+            The differentials' classes are also preserved.
+            All other matrices will change the distance, so the dimensional
+            representation is used instead.
+            Likewise, the classes of the differentials will be made dimensional.
+
+        """
+        # the transformation matrix does not need to be a rotation matrix,
+        # so the unit-distance is not guaranteed. For speed, we check if the
+        # matrix is in O(3) (rotations, proper and improper) and keeps lengths.
+        if is_rotation(matrix, allow_improper=True):  # remain in unit-rep
+            if self.differentials:
+                # TODO! shortcut if there are differentials.
+                # Currently just super, which uses Cartesian backend.
+                rep = super().transform(matrix)
+
+            else:
+                xyz = erfa_ufunc.s2c(self.lon, self.lat)
+                p = erfa_ufunc.rxp(matrix, xyz)
+                lon, lat, _ = erfa_ufunc.p2s(p)
+                rep = self.__class__(lon=lon, lat=lat)
+
+        else:  # route through dimensional representation
+            diff_cls = {k: d._dimensional_differential  # class UnitX -> X
```

Comment:
This won't work if there is already a dimensional differential attached; replace with `getattr(d, '_dimensional_differential', d.__class__)`? 

Though I guess that if there is an actual `SphericalDifferential` attached, this would fail anyway!?

Reply:
Actually, I wonder: is it better to just let it go through Cartesian (and likely fail)?

Reply:
Wouldn't routing through Cartesian propagate NaN?

Reply:
I like the ``getattr(d, '_dimensional_differential', d.__class__)``. Doing that unless...

Reply:
> Wouldn't routing through Cartesian propagate NaN (in the differential)?

I think it will regardless with the current code in `SphericalRepresentation` - since it goes through `Cartesian`...

Reply:
🤦 . Yes, of course. But when 🤞 we have representation shortcuts and transforms for the differentials...